src/libcore/result.rs

```diff
@@ -1245,6 +1245,17 @@ impl<A, E, V: FromIterator<A>> FromIterator<Result<A, E>> for Result<V, E> {
             }
         }
 
+        impl<T, E, Iter: Iterator<Item = Result<T, E>>> CollectCountGuess for Adapter<Iter, E> {
+            #[inline]
+            fn size_estimate(&self) -> usize {
+                if self.err.is_some() {
+                    0
+                } else {
+                    self.size_estimate()
```

Comment:
This should be `self.iter.size_estimate()` too; I'll check if it solves the issue shortly.